Stats: add upsell for Regions and Cities tabs in locations module

Initsogar opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Resolves https://github.com/Automattic/jetpack-roadmap/issues/2091

## Proposed Changes

* Define granular features for country, region and city tabs.
* Add upsell prompt on Region and City tabs for sites without commercial feature.
* Fix `Visitors` -> `Views`

![CleanShot 2025-01-07 at 13 36 20@2x](https://github.com/user-attachments/assets/80137712-a05e-4163-8f63-d009a16cc243)

## Demo

https://github.com/user-attachments/assets/85c1ab16-8900-42e4-9933-4c6f40aa3738

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

p1HpG7-rpL-p2

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

> [!NOTE]
> Ensure you have a test site with at least 1 visit that you could use to test it.

* Spin up a Calypso Live Branch
* Using a site with WordPress.com Personal plan, navigate to Stats page
* Enable the feature flag by appending `?flags=stats/locations`
* Look for the Locations module and you should see the Countries tab as usual
* Now try clicking on Regions or Cities
* Ensure you see the upsell prompt
* Now try using a site with WordPress.com Premium (or higher) plan
* Ensure you don't see the upsell prompt and see the data

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

Initsogar committed:
Add location views query

Initsogar committed:
Implement useLocationViewsQuery hook

Initsogar committed:
Normalize country views

Initsogar committed:
Fix country views render

Initsogar committed:
Remove 'useShortNumber' prop

Initsogar committed:
Add stats location features

Initsogar committed:
Add region and city to commercial gates

Initsogar committed:
Add upsell copy for regions and cities views

Initsogar committed:
Blur column headers when splitHeader is passed

Initsogar committed:
Add upsell on stats locations module

Initsogar committed:
Ensure tabs are positioned above the overlay

Initsogar committed:
Merge remote-tracking branch 'origin/trunk' into update/stats-locations-gate

Initsogar committed:
Add fake data

Initsogar commented on packages/components/src/horizontal-bar-list/stats-card.scss:
```diff
@@ -117,6 +117,7 @@
 		}
 
 		.stats-card__content .stats-card--header-and-body .stats-card--body,
+		.stats-card__content .stats-card-header--split .stats-card--column-header,
```
This is to blur the column headers when splitHeader is passed.

| Before | After |
|---|---|
| ![CleanShot 2025-01-07 at 13 05 19@2x](https://github.com/user-attachments/assets/64100b33-df81-4eac-a833-d394295c9bea) | ![CleanShot 2025-01-07 at 13 06 04@2x](https://github.com/user-attachments/assets/87978585-5268-4a0b-9e87-d98319c816f8) |

myhro reviewed (commented):
Hi Rafa. I think this PR needs to be updated? I'm not seeing the upsell, only the placeholder for empty data:

https://github.com/user-attachments/assets/dd721801-869a-43fc-974b-36a401ee33cf

This may be caused by the fact that we merged the gating at the API level, which is now returning `400` for `/region` and `/city` endpoints on free plans.

Can you please take a look when you get back?

kangzj commented on client/my-sites/stats/hooks/use-should-gate-stats.ts:
```diff
@@ -85,6 +87,8 @@ const jetpackStatsCommercialPaywall = [
 	STAT_TYPE_COMMENTS,
 	STATS_TYPE_DEVICE_STATS,
 	STATS_FEATURE_UTM_STATS,
+	STATS_FEATURE_LOCATION_REGION_VIEWS,
```
Feels like we don't need these feature to be here as we've got them in `granularControlForJetpackStatsCommercialPaywall` already?

Initsogar replied:
Thanks! You are right. I just removed it :)

kangzj commented on client/my-sites/stats/hooks/use-should-gate-stats.ts:
```diff
@@ -167,6 +179,8 @@ export const paidStats = [
 	STAT_TYPE_VIDEO_PLAYS,
 	STATS_TYPE_DEVICE_STATS,
 	STATS_FEATURE_UTM_STATS,
+	STATS_FEATURE_LOCATION_REGION_VIEWS,
```
We should show upsells in this case, shouldn't we?

Initsogar committed:
Merge remote-tracking branch 'origin/trunk' into update/stats-locations-gate

Initsogar committed:
Remove duplicated region and city features on stats commercial paywall config

Initsogar committed:
Merge remote-tracking branch 'origin/trunk' into update/stats-locations-gate

Initsogar committed:
Update gating by tab instead of by module

Initsogar committed:
Update locations map to render fake data

Initsogar commented:
@myhro good catch. I just addressed the comments, and it is now fixed by gating the tab rather than the whole module. Could you please re-review it?

Initsogar committed:
Update to prevent fetch location API if tab is gated

Initsogar commented on client/my-sites/stats/features/modules/stats-locations/stats-locations.tsx:
```diff
@@ -50,40 +57,46 @@ interface StatsModuleLocationsProps {
 		isError,
-	} = useLocationViewsQuery< StatsLocationViewsData >( siteId, geoMode, query );
+	} = useLocationViewsQuery< StatsLocationViewsData >( siteId, geoMode, query, {
+		enabled: ! shouldGate,
```
This is to prevent querying the location-views API if the region/city tabs are locked.

Initsogar commented:
> @myhro good catch. I just addressed the comments, and it is now fixed by gating the tab rather than the whole module. Could you please re-review it?

@myhro after this, I noted a weird bug rendering a loading placeholder on top of the locations' module. I just addressed this, and also prevented fetching the location-views API unnecessarily if we already know the region/city tab is gated. Could you please take another look? :)

Initsogar committed:
Merge remote-tracking branch 'origin/trunk' into update/stats-locations-gate

Initsogar committed:
Add region and city location as advanced paywall

myhro reviewed (approved):
Thanks, Rafa. LGTM. 👍 

The only issue is that the container might grow considerably when showing the upsell, but we can tackle that in a different issue if you prefer to merge this as is.

| Countries | Regions (locked) |
| - | - |
| <img width="1195" alt="image" src="https://github.com/user-attachments/assets/fd8d38f0-43b9-4c7e-b1b3-a1c9575287c4" /> | <img width="1195" alt="image" src="https://github.com/user-attachments/assets/2d1275fa-fff9-4e54-93f6-9d95724c2bd4" /> |

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/17197565

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @Initsogar for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/17197565).
